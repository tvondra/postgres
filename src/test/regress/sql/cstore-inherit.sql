-- create two fake cstore AM instances (no implementation)
INSERT INTO pg_cstore_am VALUES ('bar11', 1, 1, 1, 1);
INSERT INTO pg_cstore_am VALUES ('bar12', 1, 1, 1, 1);

-- parent table with two column stores
CREATE TABLE parent_table (
    a INT,
    b INT COLUMN STORE foo1 USING bar11,
    c INT,
    d INT,
    e INT,
    COLUMN STORE foo2 USING bar12 (d,e)
);

-- check contents of the catalogs
SELECT cstname, cstnatts, cstatts
  FROM pg_cstore
 WHERE cstrelid = 'parent_table'::regclass
 ORDER BY cstname;

-- basic pg_class attributes
SELECT relnamespace, reltype, reltablespace, relhasindex, relisshared, relpersistence, relkind, relnatts
  FROM pg_class WHERE oid IN (SELECT cststoreid FROM pg_cstore WHERE cstrelid = 'parent_table'::regclass)
 ORDER BY oid;

-- pg_class flags
SELECT relnatts, relchecks, relhasoids, relhaspkey, relhasrules, relhastriggers, relhassubclass, relrowsecurity, relispopulated
  FROM pg_class WHERE oid IN (SELECT cststoreid FROM pg_cstore WHERE cstrelid = 'parent_table'::regclass)
 ORDER BY oid;
 
-- pg_attribute entries (for the cstore)
SELECT cstname, attnum, attname, atttypid, attstattarget
  FROM pg_attribute JOIN pg_cstore ON (attrelid = cststoreid)
 WHERE cstrelid = 'parent_table'::regclass
 ORDER BY cstname, attnum;

-- child table with two separate column stores
CREATE TABLE child_table_1 (
    f INT,
    g INT COLUMN STORE foo1c USING bar11,
    h INT,
    i INT,
    COLUMN STORE foo2c USING bar12(h,i)
) INHERITS (parent_table);

-- check contents of the catalogs
SELECT cstname, cstnatts, cstatts
  FROM pg_cstore
 WHERE cstrelid = 'child_table_1'::regclass
 ORDER BY cstname;

-- basic pg_class attributes
SELECT relnamespace, reltype, reltablespace, relhasindex, relisshared, relpersistence, relkind, relnatts
  FROM pg_class WHERE oid IN (SELECT cststoreid FROM pg_cstore WHERE cstrelid = 'child_table_1'::regclass)
 ORDER BY oid;

-- pg_class flags
SELECT relnatts, relchecks, relhasoids, relhaspkey, relhasrules, relhastriggers, relhassubclass, relrowsecurity, relispopulated
  FROM pg_class WHERE oid IN (SELECT cststoreid FROM pg_cstore WHERE cstrelid = 'child_table_1'::regclass)
 ORDER BY oid;
 
-- pg_attribute entries (for the cstore)
SELECT cstname, attnum, attname, atttypid, attstattarget
  FROM pg_attribute JOIN pg_cstore ON (attrelid = cststoreid)
 WHERE cstrelid = 'child_table_1'::regclass
 ORDER BY cstname, attnum;

-- child table with two column stores - one modifying, one redefining the parent
CREATE TABLE child_table_2 (
    f INT,
    g INT COLUMN STORE foo1c USING bar11, -- new column store
    h INT,
    i INT,
    COLUMN STORE foo2c USING bar12(b,h,i) -- redefines the parent colstore
) INHERITS (parent_table);

-- check contents of the catalogs
SELECT cstname, cstnatts, cstatts
  FROM pg_cstore
 WHERE cstrelid = 'child_table_2'::regclass
 ORDER BY cstname;

-- basic pg_class attributes
SELECT relnamespace, reltype, reltablespace, relhasindex, relisshared, relpersistence, relkind, relnatts
  FROM pg_class WHERE oid IN (SELECT cststoreid FROM pg_cstore WHERE cstrelid = 'child_table_2'::regclass)
 ORDER BY oid;

-- pg_class flags
SELECT relnatts, relchecks, relhasoids, relhaspkey, relhasrules, relhastriggers, relhassubclass, relrowsecurity, relispopulated
  FROM pg_class WHERE oid IN (SELECT cststoreid FROM pg_cstore WHERE cstrelid = 'child_table_2'::regclass)
 ORDER BY oid;
 
-- pg_attribute entries (for the cstore)
SELECT cstname, attnum, attname, atttypid, attstattarget
  FROM pg_attribute JOIN pg_cstore ON (attrelid = cststoreid)
 WHERE cstrelid = 'child_table_2'::regclass
 ORDER BY cstname, attnum;

-- child table with a single column store of the whole table
CREATE TABLE child_table_3 (
    f INT,
    g INT,
    h INT,
    i INT,
    COLUMN STORE foo1 USING bar11(a,b,c,d,e,f,g,h,i)
) INHERITS (parent_table);

-- check contents of the catalogs
SELECT cstname, cstnatts, cstatts
  FROM pg_cstore
 WHERE cstrelid = 'child_table_3'::regclass
 ORDER BY cstname;

-- basic pg_class attributes
SELECT relnamespace, reltype, reltablespace, relhasindex, relisshared, relpersistence, relkind, relnatts
  FROM pg_class WHERE oid IN (SELECT cststoreid FROM pg_cstore WHERE cstrelid = 'child_table_3'::regclass)
 ORDER BY oid;

-- pg_class flags
SELECT relnatts, relchecks, relhasoids, relhaspkey, relhasrules, relhastriggers, relhassubclass, relrowsecurity, relispopulated
  FROM pg_class WHERE oid IN (SELECT cststoreid FROM pg_cstore WHERE cstrelid = 'child_table_3'::regclass)
 ORDER BY oid;
 
-- pg_attribute entries (for the cstore)
SELECT cstname, attnum, attname, atttypid, attstattarget
  FROM pg_attribute JOIN pg_cstore ON (attrelid = cststoreid)
 WHERE cstrelid = 'child_table_3'::regclass
 ORDER BY cstname, attnum;

-- test cleanup
CREATE TABLE cstore_oids AS
SELECT cststoreid
  FROM pg_cstore JOIN pg_class ON (pg_cstore.cstrelid = pg_class.oid)
 WHERE relname IN ('parent_table', 'child_table_1', 'child_table_2', 'child_table_3');

CREATE TABLE cstore_oids_2 AS
SELECT pg_class.oid
  FROM pg_class JOIN cstore_oids ON (pg_class.oid = cstore_oids.cststoreid);

DROP TABLE child_table_1;
DROP TABLE child_table_2;
DROP TABLE child_table_3;
DROP TABLE parent_table;

-- should return 0
SELECT COUNT(*) FROM pg_class WHERE oid IN (SELECT cststoreid FROM cstore_oids);
SELECT COUNT(*) FROM pg_class WHERE oid IN (SELECT oid FROM cstore_oids_2);

SELECT COUNT(*) FROM pg_cstore WHERE cststoreid IN (SELECT oid FROM cstore_oids);

SELECT COUNT(*) FROM pg_attribute WHERE attrelid IN (SELECT cststoreid FROM cstore_oids);
SELECT COUNT(*) FROM pg_attribute WHERE attrelid IN (SELECT oid FROM cstore_oids_2);

DROP TABLE cstore_oids;
DROP TABLE cstore_oids_2;

-- delete the fake cstore AM records
DELETE FROM pg_cstore_am;
