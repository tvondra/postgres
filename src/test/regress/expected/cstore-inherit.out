-- create two fake cstore AM instances (no implementation)
CREATE COLUMN STORE ACCESS METHOD bar11 HANDLER cstore_dummy_handler;
CREATE COLUMN STORE ACCESS METHOD bar12 HANDLER cstore_dummy_handler;
-- parent table with two column stores
CREATE TABLE parent_table (
    a INT,
    b INT COLUMN STORE foo1 USING bar11,
    c INT,
    d INT,
    e INT,
    COLUMN STORE foo2 USING bar12 (d,e)
);
-- check contents of the catalogs
SELECT relname, relhascstore, relkind
  FROM pg_class
 WHERE relname = 'parent_table';
   relname    | relhascstore | relkind 
--------------+--------------+---------
 parent_table | t            | r
(1 row)

SELECT cstname, cstnatts, cstatts
  FROM pg_cstore
 WHERE cstrelid = 'parent_table'::regclass
 ORDER BY cstname;
 cstname | cstnatts | cstatts 
---------+----------+---------
 foo1    |        1 | 2
 foo2    |        2 | 4 5
(2 rows)

-- basic pg_class attributes
SELECT relnamespace, reltype, reltablespace, relhasindex, relhascstore, relisshared, relpersistence, relkind, relnatts
  FROM pg_class WHERE oid IN (SELECT cststoreid FROM pg_cstore WHERE cstrelid = 'parent_table'::regclass)
 ORDER BY oid;
 relnamespace | reltype | reltablespace | relhasindex | relhascstore | relisshared | relpersistence | relkind | relnatts 
--------------+---------+---------------+-------------+--------------+-------------+----------------+---------+----------
            9 |       0 |             0 | f           | f            | f           | p              | C       |        2
            9 |       0 |             0 | f           | f            | f           | p              | C       |        1
(2 rows)

-- pg_class flags
SELECT relnatts, relchecks, relhasoids, relhaspkey, relhasrules, relhastriggers, relhassubclass, relrowsecurity, relispopulated
  FROM pg_class WHERE oid IN (SELECT cststoreid FROM pg_cstore WHERE cstrelid = 'parent_table'::regclass)
 ORDER BY oid;
 relnatts | relchecks | relhasoids | relhaspkey | relhasrules | relhastriggers | relhassubclass | relrowsecurity | relispopulated 
----------+-----------+------------+------------+-------------+----------------+----------------+----------------+----------------
        2 |         0 | f          | f          | f           | f              | f              | f              | t
        1 |         0 | f          | f          | f           | f              | f              | f              | t
(2 rows)

 
-- pg_attribute entries (for the cstore)
SELECT cstname, attnum, attname, atttypid, attstattarget
  FROM pg_attribute JOIN pg_cstore ON (attrelid = cststoreid)
 WHERE cstrelid = 'parent_table'::regclass
 ORDER BY cstname, attnum;
 cstname | attnum | attname | atttypid | attstattarget 
---------+--------+---------+----------+---------------
 foo1    |      1 | b       |       23 |            -1
 foo2    |      1 | d       |       23 |            -1
 foo2    |      2 | e       |       23 |            -1
(3 rows)

-- child table with two separate column stores
CREATE TABLE child_table_1 (
    f INT,
    g INT COLUMN STORE foo1c USING bar11,
    h INT,
    i INT,
    COLUMN STORE foo2c USING bar12(h,i)
) INHERITS (parent_table);
-- check contents of the catalogs
SELECT relname, relhascstore, relkind
  FROM pg_class
 WHERE relname = 'child_table_1';
    relname    | relhascstore | relkind 
---------------+--------------+---------
 child_table_1 | t            | r
(1 row)

SELECT cstname, cstnatts, cstatts
  FROM pg_cstore
 WHERE cstrelid = 'child_table_1'::regclass
 ORDER BY cstname;
 cstname | cstnatts | cstatts 
---------+----------+---------
 foo1    |        1 | 2
 foo1c   |        1 | 7
 foo2    |        2 | 4 5
 foo2c   |        2 | 8 9
(4 rows)

-- basic pg_class attributes
SELECT relnamespace, reltype, reltablespace, relhasindex, relhascstore, relisshared, relpersistence, relkind, relnatts
  FROM pg_class WHERE oid IN (SELECT cststoreid FROM pg_cstore WHERE cstrelid = 'child_table_1'::regclass)
 ORDER BY oid;
 relnamespace | reltype | reltablespace | relhasindex | relhascstore | relisshared | relpersistence | relkind | relnatts 
--------------+---------+---------------+-------------+--------------+-------------+----------------+---------+----------
            9 |       0 |             0 | f           | f            | f           | p              | C       |        2
            9 |       0 |             0 | f           | f            | f           | p              | C       |        1
            9 |       0 |             0 | f           | f            | f           | p              | C       |        2
            9 |       0 |             0 | f           | f            | f           | p              | C       |        1
(4 rows)

-- pg_class flags
SELECT relnatts, relchecks, relhasoids, relhaspkey, relhasrules, relhastriggers, relhassubclass, relrowsecurity, relispopulated
  FROM pg_class WHERE oid IN (SELECT cststoreid FROM pg_cstore WHERE cstrelid = 'child_table_1'::regclass)
 ORDER BY oid;
 relnatts | relchecks | relhasoids | relhaspkey | relhasrules | relhastriggers | relhassubclass | relrowsecurity | relispopulated 
----------+-----------+------------+------------+-------------+----------------+----------------+----------------+----------------
        2 |         0 | f          | f          | f           | f              | f              | f              | t
        1 |         0 | f          | f          | f           | f              | f              | f              | t
        2 |         0 | f          | f          | f           | f              | f              | f              | t
        1 |         0 | f          | f          | f           | f              | f              | f              | t
(4 rows)

 
-- pg_attribute entries (for the cstore)
SELECT cstname, attnum, attname, atttypid, attstattarget
  FROM pg_attribute JOIN pg_cstore ON (attrelid = cststoreid)
 WHERE cstrelid = 'child_table_1'::regclass
 ORDER BY cstname, attnum;
 cstname | attnum | attname | atttypid | attstattarget 
---------+--------+---------+----------+---------------
 foo1    |      1 | b       |       23 |            -1
 foo1c   |      1 | g       |       23 |            -1
 foo2    |      1 | d       |       23 |            -1
 foo2    |      2 | e       |       23 |            -1
 foo2c   |      1 | h       |       23 |            -1
 foo2c   |      2 | i       |       23 |            -1
(6 rows)

-- child table with two column stores - one modifying, one redefining the parent
CREATE TABLE child_table_2 (
    f INT,
    g INT COLUMN STORE foo1c USING bar11, -- new column store
    h INT,
    i INT,
    COLUMN STORE foo2c USING bar12(b,h,i) -- redefines the parent colstore
) INHERITS (parent_table);
-- check contents of the catalogs
SELECT relname, relhascstore, relkind
  FROM pg_class
 WHERE relname = 'child_table_2';
    relname    | relhascstore | relkind 
---------------+--------------+---------
 child_table_2 | t            | r
(1 row)

SELECT cstname, cstnatts, cstatts
  FROM pg_cstore
 WHERE cstrelid = 'child_table_2'::regclass
 ORDER BY cstname;
 cstname | cstnatts | cstatts 
---------+----------+---------
 foo1c   |        1 | 7
 foo2    |        2 | 4 5
 foo2c   |        3 | 2 8 9
(3 rows)

-- basic pg_class attributes
SELECT relnamespace, reltype, reltablespace, relhasindex, relhascstore, relisshared, relpersistence, relkind, relnatts
  FROM pg_class WHERE oid IN (SELECT cststoreid FROM pg_cstore WHERE cstrelid = 'child_table_2'::regclass)
 ORDER BY oid;
 relnamespace | reltype | reltablespace | relhasindex | relhascstore | relisshared | relpersistence | relkind | relnatts 
--------------+---------+---------------+-------------+--------------+-------------+----------------+---------+----------
            9 |       0 |             0 | f           | f            | f           | p              | C       |        3
            9 |       0 |             0 | f           | f            | f           | p              | C       |        1
            9 |       0 |             0 | f           | f            | f           | p              | C       |        2
(3 rows)

-- pg_class flags
SELECT relnatts, relchecks, relhasoids, relhaspkey, relhasrules, relhastriggers, relhassubclass, relrowsecurity, relispopulated
  FROM pg_class WHERE oid IN (SELECT cststoreid FROM pg_cstore WHERE cstrelid = 'child_table_2'::regclass)
 ORDER BY oid;
 relnatts | relchecks | relhasoids | relhaspkey | relhasrules | relhastriggers | relhassubclass | relrowsecurity | relispopulated 
----------+-----------+------------+------------+-------------+----------------+----------------+----------------+----------------
        3 |         0 | f          | f          | f           | f              | f              | f              | t
        1 |         0 | f          | f          | f           | f              | f              | f              | t
        2 |         0 | f          | f          | f           | f              | f              | f              | t
(3 rows)

 
-- pg_attribute entries (for the cstore)
SELECT cstname, attnum, attname, atttypid, attstattarget
  FROM pg_attribute JOIN pg_cstore ON (attrelid = cststoreid)
 WHERE cstrelid = 'child_table_2'::regclass
 ORDER BY cstname, attnum;
 cstname | attnum | attname | atttypid | attstattarget 
---------+--------+---------+----------+---------------
 foo1c   |      1 | g       |       23 |            -1
 foo2    |      1 | d       |       23 |            -1
 foo2    |      2 | e       |       23 |            -1
 foo2c   |      1 | b       |       23 |            -1
 foo2c   |      2 | h       |       23 |            -1
 foo2c   |      3 | i       |       23 |            -1
(6 rows)

-- child table with a single column store of the whole table
CREATE TABLE child_table_3 (
    f INT,
    g INT,
    h INT,
    i INT,
    COLUMN STORE foo1 USING bar11(a,b,c,d,e,f,g,h,i)
) INHERITS (parent_table);
-- check contents of the catalogs
SELECT relname, relhascstore, relkind
  FROM pg_class
 WHERE relname = 'child_table_3';
    relname    | relhascstore | relkind 
---------------+--------------+---------
 child_table_3 | t            | r
(1 row)

SELECT cstname, cstnatts, cstatts
  FROM pg_cstore
 WHERE cstrelid = 'child_table_3'::regclass
 ORDER BY cstname;
 cstname | cstnatts |      cstatts      
---------+----------+-------------------
 foo1    |        9 | 1 2 3 4 5 6 7 8 9
(1 row)

-- basic pg_class attributes
SELECT relnamespace, reltype, reltablespace, relhasindex, relhascstore, relisshared, relpersistence, relkind, relnatts
  FROM pg_class WHERE oid IN (SELECT cststoreid FROM pg_cstore WHERE cstrelid = 'child_table_3'::regclass)
 ORDER BY oid;
 relnamespace | reltype | reltablespace | relhasindex | relhascstore | relisshared | relpersistence | relkind | relnatts 
--------------+---------+---------------+-------------+--------------+-------------+----------------+---------+----------
            9 |       0 |             0 | f           | f            | f           | p              | C       |        9
(1 row)

-- pg_class flags
SELECT relnatts, relchecks, relhasoids, relhaspkey, relhasrules, relhastriggers, relhassubclass, relrowsecurity, relispopulated
  FROM pg_class WHERE oid IN (SELECT cststoreid FROM pg_cstore WHERE cstrelid = 'child_table_3'::regclass)
 ORDER BY oid;
 relnatts | relchecks | relhasoids | relhaspkey | relhasrules | relhastriggers | relhassubclass | relrowsecurity | relispopulated 
----------+-----------+------------+------------+-------------+----------------+----------------+----------------+----------------
        9 |         0 | f          | f          | f           | f              | f              | f              | t
(1 row)

 
-- pg_attribute entries (for the cstore)
SELECT cstname, attnum, attname, atttypid, attstattarget
  FROM pg_attribute JOIN pg_cstore ON (attrelid = cststoreid)
 WHERE cstrelid = 'child_table_3'::regclass
 ORDER BY cstname, attnum;
 cstname | attnum | attname | atttypid | attstattarget 
---------+--------+---------+----------+---------------
 foo1    |      1 | a       |       23 |            -1
 foo1    |      2 | b       |       23 |            -1
 foo1    |      3 | c       |       23 |            -1
 foo1    |      4 | d       |       23 |            -1
 foo1    |      5 | e       |       23 |            -1
 foo1    |      6 | f       |       23 |            -1
 foo1    |      7 | g       |       23 |            -1
 foo1    |      8 | h       |       23 |            -1
 foo1    |      9 | i       |       23 |            -1
(9 rows)

-- test cleanup
CREATE TABLE cstore_oids AS
SELECT cststoreid
  FROM pg_cstore JOIN pg_class ON (pg_cstore.cstrelid = pg_class.oid)
 WHERE relname IN ('parent_table', 'child_table_1', 'child_table_2', 'child_table_3');
CREATE TABLE cstore_oids_2 AS
SELECT pg_class.oid
  FROM pg_class JOIN cstore_oids ON (pg_class.oid = cstore_oids.cststoreid);
DROP TABLE child_table_1;
DROP TABLE child_table_2;
DROP TABLE child_table_3;
DROP TABLE parent_table;
-- should return 0
SELECT COUNT(*) FROM pg_class WHERE oid IN (SELECT cststoreid FROM cstore_oids);
 count 
-------
     0
(1 row)

SELECT COUNT(*) FROM pg_class WHERE oid IN (SELECT oid FROM cstore_oids_2);
 count 
-------
     0
(1 row)

SELECT COUNT(*) FROM pg_cstore WHERE cststoreid IN (SELECT oid FROM cstore_oids);
 count 
-------
     0
(1 row)

SELECT COUNT(*) FROM pg_attribute WHERE attrelid IN (SELECT cststoreid FROM cstore_oids);
 count 
-------
     0
(1 row)

SELECT COUNT(*) FROM pg_attribute WHERE attrelid IN (SELECT oid FROM cstore_oids_2);
 count 
-------
     0
(1 row)

DROP TABLE cstore_oids;
DROP TABLE cstore_oids_2;
-- delete the fake cstore AM records
DELETE FROM pg_cstore_am;
